src/main/groovy/life/qbic/portal/sampletracking/datasources/SampleTracker.groovy

```diff
@@ -78,23 +79,33 @@ class SampleTracker {
             URI locationsUri = new URI("${service.rootUrl.toExternalForm()}/locations/$emailAddress")
 
             HttpRequest request = HttpRequest.GET(locationsUri).basicAuth(user.name, user.password)
-            HttpResponse<List> response
 
-            client.withCloseable {
-                response = it.toBlocking().exchange(request)
-            }
+            HttpResponse<?> response
 
-            if (response?.status?.code != 200) {
-                throw new SampleTrackingQueryException("Request for current location failed.")
+            try {
+                response = client.withCloseable { it.toBlocking().exchange(request,
+                                                    Argument.of(List.class, Location.class))}
+            }
+            catch(HttpClientResponseException e) {
+                response = e.response
+                log.error("Response code was greater or equal to 400.", e)
+                if (response?.status?.code == 400) {
+                    throw new SampleTrackingQueryException("Invalid email $emailAddress requested.")
+                } else if (response?.status?.code == 404) {
+                    throw new SampleTrackingQueryException("Location for requested email address $emailAddress  could not be found.")
+                } else if (response?.status?.code != 200) {
+                    throw new SampleTrackingQueryException("Request for current location failed.")
+                }
             }
-            if (!response?.body() instanceof List) {
-                throw new SampleTrackingQueryException("Did not receive a valid List response.")
+
+            if (response.getBody().empty()) {
```

Comment:
Quite the improvement. Way more readable. We should do it like this in every comparable method.